File: packages/negative-support/src/license.ts
/**
 * License management for negative-support (Node.js CLI).
 *
 * Shares config files with the Python CLI at ~/.negative-support/.
 * Requires a valid ns_live_* token to run. Tokens are issued when you
 * sign in at https://negative.support. Free tokens get 10 runs, paid
 * tokens get unlimited.
 */

import { createHash } from 'crypto';
import { hostname, platform, arch, networkInterfaces, userInfo } from 'os';
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'fs';
import { join } from 'path';
import { homedir } from 'os';

// ── Configuration ─────────────────────────────────────────────────────

const FREE_RUNS = 10;
const TOKEN_PREFIX = 'ns_live_';
const API_BASE = process.env.NS_API_BASE || 'https://negative.support';
const GRACE_DAYS = 7;
const CONFIG_DIR = join(homedir(), '.negative-support');
const LICENSE_FILE = join(CONFIG_DIR, 'license.json');

// ── Machine fingerprint ───────────────────────────────────────────────

function getMachineId(): string {
  // Match Python: SHA-256 of hostname|MAC|OS|arch|username
  const mac = getMacAddress();
  const parts = [
    hostname(),
    mac,
    platform(),
    arch(),
    userInfo().username,
  ];
  return createHash('sha256').update(parts.join('|')).digest('hex');
}

function getMacAddress(): string {
  // Match Python's uuid.getnode() — returns MAC as decimal integer string
  const ifaces = networkInterfaces();
  for (const name of Object.keys(ifaces)) {
    for (const iface of ifaces[name]!) {
      if (!iface.internal && iface.mac && iface.mac !== '00:00:00:00:00:00') {
        // Convert MAC to integer like Python's uuid.getnode()
        const hex = iface.mac.replace(/:/g, '');
        return BigInt('0x' + hex).toString();
      }
    }
  }
  return '0';
}

// ── Local storage ─────────────────────────────────────────────────────

function ensureConfigDir(): void {
  if (!existsSync(CONFIG_DIR)) mkdirSync(CONFIG_DIR, { recursive: true });
}

function readJson(path: string): Record<string, unknown> | null {
  try {
    return JSON.parse(readFileSync(path, 'utf-8'));
  } catch {
    return null;
  }
}

function writeJson(path: string, data: Record<string, unknown>): void {
  ensureConfigDir();
  writeFileSync(path, JSON.stringify(data, null, 2) + '\n');
}

// ── Server communication ─────────────────────────────────────────────

async function apiPost(endpoint: string, body: Record<string, unknown>): Promise<Record<string, unknown> | null> {
  const url = `${API_BASE}${endpoint}`;
  try {
    const resp = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(5000),
    });
    return await resp.json() as Record<string, unknown>;
  } catch {
    return null;
  }
}

// ── Token validation ─────────────────────────────────────────────────

function isValidTokenFormat(token: string): boolean {
  if (!token.startsWith(TOKEN_PREFIX)) return false;
  const hex = token.slice(TOKEN_PREFIX.length);
  return hex.length === 32 && /^[0-9a-f]+$/.test(hex);
}

async function validateLicense(): Promise<[boolean, string]> {
  const lic = readJson(LICENSE_FILE);
  if (!lic || typeof lic.token !== 'string') return [false, ''];

  const token = lic.token as string;
  if (!isValidTokenFormat(token)) return [false, 'Invalid token format.'];

  // Try server validation
  const resp = await apiPost('/api/validate', { token });
  if (resp !== null) {
    if (resp.valid) {
      lic.last_validated = new Date().toISOString();
      lic.plan = (resp.plan as string) || 'lifetime';
      lic.runs_used = resp.runs_used ?? 0;
      writeJson(LICENSE_FILE, lic);
      const plan = lic.plan as string;
      if (plan === 'lifetime') {
        return [true, 'Licensed (lifetime)'];
      } else {
        const remaining = (resp.free_remaining as number) ?? (FREE_RUNS - ((resp.runs_used as number) ?? 0));
        return [true, `Free tier (${remaining} run${remaining !== 1 ? 's' : ''} remaining)`];
      }
    } else {
      const error = (resp.error as string) || 'Token is no longer valid.';
      if (error.toLowerCase().includes('exhausted') || ((resp.runs_used as number) ?? 0) >= FREE_RUNS) {
        return [false, 'exhausted'];
      }
      return [false, error];
    }
  }

  // Server unreachable — check grace period
  const lastValidated = lic.last_validated as string | undefined;
  if (lastValidated) {
    try {
      const lastDt = new Date(lastValidated);
      const daysOffline = Math.floor((Date.now() - lastDt.getTime()) / 86400000);
      if (daysOffline <= GRACE_DAYS) {
        return [true, `Licensed (offline, validated ${daysOffline}d ago)`];
      } else {
        return [false, `License not validated in ${daysOffline} days. Connect to the internet to re-validate.`];
      }
    } catch { /* fall through */ }
  }

  return [false, 'Cannot validate license (server unreachable).'];
}

// ── Public API ────────────────────────────────────────────────────────

export async function checkLicense(): Promise<[boolean, string]> {
  const [hasLicense, msg] = await validateLicense();
  if (hasLicense) return [true, msg];

  if (msg === 'exhausted') return [false, 'exhausted'];

  // No token found
  return [false, 'no_token'];
}

export async function activateToken(token: string): Promise<[boolean, string]> {
  token = token.trim();
  if (!isValidTokenFormat(token)) {
    return [false, `Invalid token format. Expected: ${TOKEN_PREFIX}<32 hex characters>\nExample: ${TOKEN_PREFIX}${'a1b2c3d4'.repeat(4)}`];
  }

  const machineId = getMachineId();
  const resp = await apiPost('/api/activate', { token, machine_id: machineId });
  const now = new Date().toISOString();
  const lic: Record<string, unknown> = {
    token,
    activated_at: now,
    last_validated: now,
    plan: 'lifetime',
  };

  if (resp !== null) {
    if (resp.ok) {
      lic.plan = (resp.plan as string) || 'lifetime';
      writeJson(LICENSE_FILE, lic);
      return [true, `License activated! Plan: ${lic.plan}`];
    } else {
      return [false, (resp.error as string) || 'Server rejected this token.'];
    }
  }

  // Server unreachable — save locally
  writeJson(LICENSE_FILE, lic);
  return [true, 'Token saved. Could not reach server for validation — it will be verified on next online run.'];
}

export async function getStatus(): Promise<string> {
  const lines: string[] = [];
  const lic = readJson(LICENSE_FILE);

  if (lic && typeof lic.token === 'string') {
    const token = lic.token as string;
    const masked = token.length > 12 ? token.slice(0, 8) + '...' + token.slice(-4) : token;
    lines.push(`Token:      ${masked}`);
    lines.push(`Plan:       ${lic.plan || 'unknown'}`);
    lines.push(`Activated:  ${lic.activated_at || 'unknown'}`);
    lines.push(`Validated:  ${lic.last_validated || 'never'}`);

    const [valid, msg] = await validateLicense();
    lines.push(`Status:     ${valid ? 'valid' : 'invalid'} — ${msg}`);
  } else {
    lines.push('Token:      none');
    lines.push('');
    lines.push('Sign in at https://negative.support to get your token,');
    lines.push('then activate it:');
    lines.push('  negative-support --activate <your-token>');
  }

  return lines.join('\n');
}

export function printNoTokenMessage(): void {
  console.log();
  console.log('  No license token found.');
  console.log();
  console.log('  1. Sign in at https://negative.support');
  console.log('  2. Copy your token from the user menu');
  console.log('  3. Run: negative-support --activate <your-token>');
  console.log();
  console.log('  Free accounts get 10 runs. Buy a lifetime license for unlimited use.');
  console.log();
}

export function printExhaustedMessage(): void {
  console.log();
  console.log(`  Free tier exhausted (${FREE_RUNS}/${FREE_RUNS} runs used).`);
  console.log();
  console.log('  To continue, buy a lifetime license at:');
  console.log('    https://negative.support/#pricing');
  console.log();
  console.log('  Your existing token will be upgraded — no re-activation needed.');
  console.log();
}

File: src/negative_support/license.py
"""License management for negative-support.

Requires a valid license token (ns_live_*) to run. Tokens are issued
when you sign in at https://negative.support. Free tokens get 10 runs,
paid tokens get unlimited.
"""

from __future__ import annotations

import datetime
import getpass
import hashlib
import json
import os
import platform
import sys
import uuid
from pathlib import Path
from urllib import request, error as urlerror

# ── Configuration ─────────────────────────────────────────────────────

FREE_RUNS = 10
CONFIG_DIR = Path.home() / ".negative-support"
LICENSE_FILE = CONFIG_DIR / "license.json"
TOKEN_PREFIX = "ns_live_"
API_BASE = os.environ.get("NS_API_BASE", "https://negative.support")
GRACE_DAYS = 7  # allow offline usage for this many days after last validation


# ── Machine fingerprint ───────────────────────────────────────────────

def _get_machine_id() -> str:
    """Generate a stable machine fingerprint.

    SHA-256 of hostname + MAC address + OS + username.
    Survives reinstalls but changes if user switches machines.
    """
    parts = [
        platform.node(),           # hostname
        str(uuid.getnode()),       # MAC address as int
        platform.system(),         # OS
        platform.machine(),        # architecture
        getpass.getuser(),         # username
    ]
    raw = "|".join(parts)
    return hashlib.sha256(raw.encode()).hexdigest()


# ── Local storage ─────────────────────────────────────────────────────

def _ensure_config_dir() -> None:
    CONFIG_DIR.mkdir(parents=True, exist_ok=True)


def _read_json(path: Path) -> dict | None:
    try:
        return json.loads(path.read_text())
    except (FileNotFoundError, json.JSONDecodeError):
        return None


def _write_json(path: Path, data: dict) -> None:
    _ensure_config_dir()
    path.write_text(json.dumps(data, indent=2) + "\n")


# ── Server communication ─────────────────────────────────────────────

def _api_post(endpoint: str, body: dict, timeout: float = 5.0) -> dict | None:
    """POST JSON to the license server. Returns parsed response or None on failure."""
    url = f"{API_BASE}{endpoint}"
    data = json.dumps(body).encode()
    req = request.Request(
        url,
        data=data,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        with request.urlopen(req, timeout=timeout) as resp:
            return json.loads(resp.read())
    except (urlerror.URLError, OSError, json.JSONDecodeError, TimeoutError):
        return None


# ── Token validation ─────────────────────────────────────────────────

def _is_valid_token_format(token: str) -> bool:
    """Check if token matches expected format: ns_live_<32 hex chars>."""
    if not token.startswith(TOKEN_PREFIX):
        return False
    hex_part = token[len(TOKEN_PREFIX):]
    return len(hex_part) == 32 and all(c in "0123456789abcdef" for c in hex_part)


def _validate_token_server(token: str) -> dict | None:
    """Validate token with server. Returns response or None if unreachable."""
    return _api_post("/api/validate", {"token": token})


def _validate_license() -> tuple[bool, str]:
    """Check if a valid license exists.

    Returns (is_valid, message).
    """
    lic = _read_json(LICENSE_FILE)
    if lic is None or "token" not in lic:
        return False, ""

    token = lic["token"]
    if not _is_valid_token_format(token):
        return False, "Invalid token format."

    # Try server validation
    resp = _validate_token_server(token)
    if resp is not None:
        if resp.get("valid"):
            # Update last validated timestamp
            lic["last_validated"] = datetime.datetime.now(datetime.timezone.utc).isoformat()
            lic["plan"] = resp.get("plan", "lifetime")
            lic["runs_used"] = resp.get("runs_used", 0)
            _write_json(LICENSE_FILE, lic)
            plan = lic["plan"]
            if plan == "lifetime":
                return True, "Licensed (lifetime)"
            else:
                remaining = resp.get("free_remaining", FREE_RUNS - resp.get("runs_used", 0))
                return True, f"Free tier ({remaining} run{'s' if remaining != 1 else ''} remaining)"
        else:
            error = resp.get("error", "Token is no longer valid.")
            if "exhausted" in str(error).lower() or resp.get("runs_used", 0) >= FREE_RUNS:
                return False, "exhausted"
            return False, error

    # Server unreachable — check grace period
    last_validated = lic.get("last_validated")
    if last_validated:
        try:
            last_dt = datetime.datetime.fromisoformat(last_validated)
            now = datetime.datetime.now(datetime.timezone.utc)
            days_offline = (now - last_dt).days
            if days_offline <= GRACE_DAYS:
                return True, f"Licensed (offline, validated {days_offline}d ago)"
            else:
                return False, (
                    f"License not validated in {days_offline} days. "
                    "Connect to the internet to re-validate."
                )
        except (ValueError, TypeError):
            pass

    return False, "Cannot validate license (server unreachable)."


# ── Public API ────────────────────────────────────────────────────────

def check_license() -> tuple[bool, str]:
    """Check if the user is allowed to run.

    Returns (allowed, message).
    Requires a valid ns_live_* token. No anonymous free tier.
    """
    has_license, msg = _validate_license()
    if has_license:
        return True, msg

    if msg == "exhausted":
        return False, "exhausted"

    # No token found
    return False, "no_token"


def activate_token(token: str) -> tuple[bool, str]:
    """Activate a license token.

    Returns (success, message).
    """
    token = token.strip()
    if not _is_valid_token_format(token):
        return False, (
            f"Invalid token format. Expected: {TOKEN_PREFIX}<32 hex characters>\n"
            f"Example: {TOKEN_PREFIX}{'a1b2c3d4' * 4}"
        )

    # Try server activation
    machine_id = _get_machine_id()
    resp = _api_post("/api/activate", {"token": token, "machine_id": machine_id})

    now = datetime.datetime.now(datetime.timezone.utc).isoformat()
    lic = {
        "token": token,
        "activated_at": now,
        "last_validated": now,
        "plan": "lifetime",
    }

    if resp is not None:
        if resp.get("ok"):
            lic["plan"] = resp.get("plan", "lifetime")
            _write_json(LICENSE_FILE, lic)
            return True, f"License activated! Plan: {lic['plan']}"
        else:
            return False, resp.get("error", "Server rejected this token.")
    else:
        # Server unreachable — save locally, validate later
        _write_json(LICENSE_FILE, lic)
        return True, (
            "Token saved. Could not reach server for validation — "
            "it will be verified on next online run."
        )


def get_status() -> str:
    """Get a human-readable license status string."""
    lines = []

    lic = _read_json(LICENSE_FILE)
    if lic and "token" in lic:
        token = lic["token"]
        masked = token[:8] + "..." + token[-4:] if len(token) > 12 else token
        lines.append(f"Token:      {masked}")
        lines.append(f"Plan:       {lic.get('plan', 'unknown')}")
        lines.append(f"Activated:  {lic.get('activated_at', 'unknown')}")
        lines.append(f"Validated:  {lic.get('last_validated', 'never')}")

        valid, msg = _validate_license()
        lines.append(f"Status:     {'valid' if valid else 'invalid'} — {msg}")
    else:
        lines.append("Token:      none")
        lines.append("")
        lines.append("Sign in at https://negative.support to get your token,")
        lines.append("then activate it:")
        lines.append("  negative-support --activate <your-token>")

    return "\n".join(lines)


def print_no_token_message() -> None:
    """Print the message shown when no token is configured."""
    print()
    print("  No license token found.")
    print()
    print("  1. Sign in at https://negative.support")
    print("  2. Copy your token from the user menu")
    print("  3. Run: negative-support --activate <your-token>")
    print()
    print("  Free accounts get 10 runs. Buy a lifetime license for unlimited use.")
    print()


def print_exhausted_message() -> None:
    """Print the message shown when free runs are exhausted."""
    print()
    print(f"  Free tier exhausted ({FREE_RUNS}/{FREE_RUNS} runs used).")
    print()
    print("  To continue, buy a lifetime license at:")
    print("    https://negative.support/#pricing")
    print()
    print("  Your existing token will be upgraded — no re-activation needed.")
    print()
